### ISSUE TITLE
Find applications for non-metallic epsilon

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4758]
**TL;DR:** The P3M algorithm provides an undocumented `epsilon` argument that modifies forces and energies.

The goal is to find physical applications of the epsilon term, and better document its role in long-range calculations.

In addition, one should look into the definition of the electric dipole moment in ESPResSo, which currently sums over unfolded coordinates. As a result, purposefully moving a particle from the unit cell to its position in a periodic image of the cell changes the orientation and magnitude of the net electric dipole moment, and artificially changes calculated energies and forces.

## Electrostatics

Here is a summary of the term involved in electrostatics:

| term | equation |
| ------- | -------------- |
| electrostatic prefactor | $\displaystyle C = \frac{1}{4\pi \varepsilon_0 \varepsilon_r}$ |
| electric dipole moment | $\vec{p} = \sum_i q_i \vec{r}_i$ |
| force correction | $\displaystyle \vec{f}_i^{\text{corr}} = - C \frac{4\pi}{V}\frac{q_i\vec{p}}{1 + 2 \epsilon}$ |
| energy correction | $\displaystyle E^{\text{corr}} = C \frac{4\pi}{V}\frac{\lVert\vec{p}\rVert}{1 + 2 \epsilon}$ |

The expression of $E^{\text{corr}}$ can be found in the second line of eq. 8 in Makov et Payne 1995, along with an illustration of the issue with the choice of supercell in fig. 1. For the special case of a cubic lattice in a cubic supercell, dipoles don't interact and the correction term vanishes. Kantorovich 1999 proposed a generalization for non-cubic supercells.

The correction is tested in the `test_p3m()` function of the `testsuite/python/electrostatic_interactions.py` test case.

The force correction calculation is carried out in the core as follows:
https://github.com/espressomd/espresso/blob/6a56470d3371ec04940cc139d750d27ac86df1ba/src/core/electrostatics/p3m.cpp#L467-L468
https://github.com/espressomd/espresso/blob/6a56470d3371ec04940cc139d750d27ac86df1ba/src/core/electrostatics/p3m.cpp#L516-L522

## Magnetostatics

Here is a summary of the term involved in magnetostatics:

| term | equation |
| ------- | -------------- |
| magnetostatic prefactor | $\displaystyle D = \frac{\mu_0 \mu_r}{4\pi}$ |
| magnetic dipole moment | $\vec{m} = \sum_i \vec{\mu}_i$ |
| energy correction | $\displaystyle E^{\text{corr}} = D \frac{4\pi}{2V}\frac{\lVert\vec{m}\rVert}{1 + 2 \epsilon}$ |
| torque correction | $\displaystyle \vec{\tau}_i^{\text{corr}} = - D \frac{4\pi}{V}\frac{\vec{\mu}_i \times \vec{m}}{1 + 2 \epsilon}$ |
| force correction | $\displaystyle \vec{f}_i^{\text{corr}} = 0$ |

The correction is tested in the `test_dp3m()` function of the `testsuite/python/dipolar_p3m.py` test case.

The force correction calculation is carried out in the core as follows:
https://github.com/espressomd/espresso/blob/6a56470d3371ec04940cc139d750d27ac86df1ba/src/core/magnetostatics/dp3m.cpp#L506-L507
https://github.com/espressomd/espresso/blob/6a56470d3371ec04940cc139d750d27ac86df1ba/src/core/magnetostatics/dp3m.cpp#L551-L555
https://github.com/espressomd/espresso/blob/6a56470d3371ec04940cc139d750d27ac86df1ba/src/core/magnetostatics/dp3m.cpp#L558-L564

## Literature

* Martin, Comment on calculations of electric polarization in crystals, *Physical Review B*, **1974**, 9(4):1998-1999, [doi:10.1103/physrevb.9.1998](https://doi.org/10.1103/physrevb.9.1998).
* Makov et Payne, Periodic boundary conditions in *ab initio* calculations, *Physical Review B*, **1995**, 51(7):4014-4022, [doi:10.1103/physrevb.51.4014](https://doi.org/10.1103/physrevb.51.4014).
* Kantorovich, Elimination of the long-range dipole interaction in calculations with periodic boundary conditions, *Physical Review B*, **1999**, 60(23):15476-15479, [doi:10.1103/physrevb.60.15476](https://doi.org/10.1103/physrevb.60.15476).
* Kantorovich et Tupitsyn, Coulomb potential inside a large finite crystal, *Journal of Physics: Condensed Matter*, **1999**, 11(32):6159-6168, [doi:10.1088/0953-8984/11/32/307](https://doi.org/10.1088/0953-8984/11/32/307).
* Ditler, Mattiat, Luber, The position operator problem in periodic calculations with an emphasis on theoretical spectroscopy, *Physical Chemistry Chemical Physics*, **2023**, 25(21):14672-14685, [doi:10.1039/d2cp05991f](https://doi.org/10.1039/d2cp05991f).
* Rutter, Periodic boundary conditions and ambiguous dipoles, *C2x*, https://www.c2x.org.uk/pbc_dipoles.html
* Rutter, Periodic boundary conditions and dipoles, *C2x*, https://www.c2x.org.uk/c2x_dipoles.html
